cli/azd/docs/external-prompting.md

```diff
@@ -0,0 +1,114 @@
+
+- `Prompt`: Asks the user for a single (freeform) text value.
+- `Select`: Asks the user to pick a single value from a list of options.
+- `MulitSelect`: Asks the user to pick zero or more values from a list of options.
+- `Confirm`: Asks the user to confirm and operation with a given message.
+
+The implementation of this interface uses a go library to provide a terminal experience (using ANSI escape sequences to provide a nice terminal interaction model) with a fallback to raw text input when the user is not connected to a proper terminal.
+
+This is a reasonable experience for users interacting with `azd` via their terminal.  However, `azd` also supports being used within IDEs (today Visual Studio Code, tomorrow Visual Studio as well) and there our terminal based prompting strategy is not ideal. VS Code is forced to run us in an interactive terminal and the user has to interact with `azd` via its terminal interface or specifically craft their calls of `azd` to not prompt the user.  In Visual Studio, AZD is run in a background process, so no terminal interaction is possible.
+
+In both cases it would be ideal if `azd` could delegate the prompting behavior back to the caller.  This document outlines a solution, which provides a way for an external tool to provide a remote service that `azd` interacts with when it needs to prompt the user for information.
+
+## Solution
+
+Similar to our strategy for delegating authentication to an external host, we support delegating prompting to an external host via a special JSON based REST API, hosted over a local HTTP server.  When run, `azd` looks for two special environment variables:
+
+- `AZD_UI_PROMPT_ENDPOINT`
+- `AZD_UI_PROMPT_KEY`
+
+When both are set, instead of prompting using the command line - the implementation of our prompting methods now make a POST call to a special endpoint:
+
+`${AZD_UI_PROMPT_ENDPOINT}/prompt?api-version=2024-02-14-preview`
+
+Setting the following headers:
+
+- `Content-Type: application/json`
+- `Authorization: Bearer ${AZD_UI_PROMPT_KEY}`
+
+The use of `AZD_UI_PROMPT_KEY` allows the host to block requests coming from other clients on the same machine (since the it is expected the host runs a ephemeral HTTP server listing on `127.0.0.1` on a random port). It is expected that the host will generate a random string and use this as a shared key for the lifetime of an `azd` invocation.
+
+The body of the request contains information about the prompt that `azd` needs a response for:
+
+```jsonc
+{
+    "type": "<string>", // one of "string", "password", "select", "multiSelect", "confirm",
+    "options": {
+      "message": "<string>", // the message to be displayed as part of the prompt
+      "help": "<string>", // optional help text that can be displayed upon request
+      "options": [ "<string>" /* ... */  ], // for select and multiSelect types - the valid options the user should pick from
```

Comment:
Might want to have an `optionsData` field of some sort, so you can have separate display names from option data without having to reparse the display name later. For example, something like this:

```jsonc
// ...
"options": "Subscription A", "Subscription B", "Subscription C"]
"optionsData": ["AAAA-AAAA...", "BBBB-BBBB...", "CCCC-CCCC..."]
```

And have the result be the subscription ID `"AAAA-AAAA..."` instead of the display name.

Reply:
The other thought that I had here would be that for an `options` request the response could be either an array of integers or an array of strings where each string is in an integer (if we wanted to stay in pure string land) which is the index of the selected options that the user picked.

Reply:
Going to mark this resolved for now, we can add if we need it from VSCode side